### PR TITLE
feat: IonConnect seal and gift wrap services

### PR DESCRIPTION
**Pull request by ice-kreios**
## Description
This PR add seal and gift wrap services to encode/decode ion-connect events.

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Refactoring
- [ ] Documentation
- [ ] Chore

### CODE REVIEW

File: lib/app/services/nostr/nostr_seal_service.dart [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import 'package:ion/app/services/nostr/ed25519_key_store.dart';
+import 'package:ion/app/utils/date.dart';
+import 'package:nip44/nip44.dart';
+import 'package:nostr_dart/nostr_dart.dart';
+
+class NostrSealService {
+  /// https://github.com/nostr-protocol/nips/blob/master/59.md#2-seal-the-rumor
+  static const int kind = 13;
+
+  Future<EventMessage> encode(EventMessage inputEvent) async {
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_seal_service.dart]: fixed

File: lib/app/services/nostr/nostr_seal_service.dart [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    final keyStore = await Ed25519KeyStore.generate();
+
+    final rumor = jsonEncode(inputEvent.toJson());
+
+    final seal = await Nip44.encryptMessage(
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_seal_service.dart]: fixed

File: lib/app/exceptions/exceptions.dart [hunk 3]
```diff
@@ -163,3 +163,7 @@ class CompressWithBrotliException extends IONException {
 class DecompressBrotliException extends IONException {
   DecompressBrotliException() : super(10034, 'Failed to decompress Brotli file.');
 }
+
+class ReceiverPubkeyNotFoundException extends IONException {
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/exceptions/exceptions.dart]: deleted

File: lib/app/services/nostr/nostr_seal_service.dart [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      kind: sealKind,
+      createdAt: createdAt,
+      content: encryptedRumor,
+    );
+  }
+
+  @override
+  Future<EventMessage> decodeSeal(
+    EventMessage seal,
+    EventSigner recipientSigner,
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_seal_service.dart]: renamed

File: lib/app/services/nostr/nostr_seal_service.dart [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+class NostrSealServiceImpl implements NostrSealService {
+  static const int sealKind = 13;
+
+  @override
+  Future<EventMessage> createSeal(
+    EventMessage rumor,
+    EventSigner signer,
+    String recipientPublicKey,
+  ) async {
+    final encodedRumor = jsonEncode(rumor.toJson());
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_seal_service.dart:33]: https://github.com/nostr-protocol/nips/blob/master/17.md#encrypting

i think we should encrypt whole List https://github.com/nostr-protocol/nips/blob/master/17.md#encrypting
In this doc, they wrap the rumor with `nip44Encrypt`

File: lib/app/services/nostr/nostr_gift_wrap_service.dart [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    String recipientPublicKey,
+  );
+
+  Future<EventMessage> decodeWrap(
+    EventMessage wrap,
+    EventSigner recipientSigner,
+  );
+}
+
+class NostrGiftWrapServiceImpl implements NostrGiftWrapService {
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_gift_wrap_service.dart:25]: yes, also added unit tests

File: lib/app/services/nostr/nostr_gift_wrap_service.dart [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  @override
+  Future<EventMessage> decodeWrap(
+    EventMessage wrap,
+    EventSigner recipientSigner,
+  ) async {
+    final decryptedContent = await Nip44.decryptMessage(
+      wrap.content,
+      recipientSigner.privateKey,
+      wrap.tags.firstWhere((tag) => tag[0] == RelatedPubkey.tagName)[1],
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_gift_wrap_service.dart]: refactored

File: lib/app/services/nostr/nostr_gift_wrap_service.dart [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+  @override
+  Future<EventMessage> createWrap(
+    EventMessage event,
+    String recipientPublicKey,
+  ) async {
+    final randomSigner = KeyStore.generate();
+
+    final encryptedEvent = await Nip44.encryptMessage(
+      jsonEncode(event.toJson()),
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_gift_wrap_service.dart:35]: https://github.com/nostr-protocol/nips/blob/master/17.md#encrypting

i think we should encrypt whole List https://github.com/nostr-protocol/nips/blob/master/17.md#encrypting
In this doc, they wrap the rumor with nip44Encrypt

[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/services/nostr/nostr_gift_wrap_service.dart]: fixed

File: test/app/services/nostr/nostr_gift_wrap_service_test.dart [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import 'package:nostr_dart/nostr_dart.dart';
+
+void main() {
+  late NostrGiftWrapService giftWrapService;
+  late EventSigner signer;
+  const pubkey = 'c95c07ad5aad2d81a3890f13b3eaa80a3d8aca173a91dc2be9fd04720a5a9377';
+
+  setUp(() {
+    giftWrapService = NostrGiftWrapServiceImpl();
+    signer = KeyStore.generate();
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply test/app/services/nostr/nostr_gift_wrap_service_test.dart]: updated

File: test/app/services/nostr/nostr_seal_service_test.dart [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import 'package:nostr_dart/nostr_dart.dart';
+
+void main() {
+  late NostrSealService sealService;
+  late EventSigner signer;
+  const pubkey = 'c95c07ad5aad2d81a3890f13b3eaa80a3d8aca173a91dc2be9fd04720a5a9377';
+
+  setUp(() {
+    sealService = NostrSealServiceImpl();
+    signer = KeyStore.generate();
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply test/app/services/nostr/nostr_seal_service_test.dart]: done

File: test/app/services/nostr/nostr_gift_wrap_service_test.dart [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  });
+
+  group('NostrGiftWrapService', () {
+    test('creates wrap from event', () async {
+      final event = await EventMessage.fromData(
```
[the remark this answers is not in the document]
Comment (ice-kreios) [reply test/app/services/nostr/nostr_gift_wrap_service_test.dart]: fixed

**Review by ice-alcides — APPROVED** [review]

**Pull request merged**